integration_tests/src/main/python/data_gen.py

```diff
@@ -208,6 +209,37 @@ def __init__(self, nullable=True, min_val = INT_MIN, max_val = INT_MAX,
     def start(self, rand):
         self._start(rand, lambda : rand.randint(self._min_val, self._max_val))
 
+
+#Maximum number of decimal digits a Long can represent is 18.
+DECIMAL_MIN = -(1 << 59)
+DECIMAL_MAX = (1 << 59)  #approximate
+class DecimalGen(DataGen):
+    """Generate Decimals, with some built in corner cases."""
+    def __init__(self, precision=None, scale=None, nullable=True, special_cases=[]):
+        if precision is None:
+            precision = 18
+            scale = 0
+            special_cases = [Decimal('-1'), Decimal('0'), Decimal('1'), Decimal(DECIMAL_MIN), Decimal(DECIMAL_MAX)]
```

Comment:
Even -1 and 1 do not work all of the time.  a precision of 1 and scale of -3 would be an example where this would not work. It can only hold multiples of 100 up to 900.

Reply:
Sorry, I don't understand above statement.  precision of 1 and scale of -3 is `1000`  and I am not getting why it can hold multiples of 100. Could you please explain. I tried setting precision and scale as mentioned above and it passed. So I am missing something.

Reply:
```
>>> df = spark.createDataFrame(data=[(Decimal('-1')), (Decimal('0'))], schema=DecimalType(7, -3))
>>> df.show()
+-----+
|value|
+-----+
| 0E+3|
| 0E+3|
+-----+
```

The details of -2 vs -3 scale don't matter, what matters is -1 becomes 0 for any scale that is negative.  For a Decimal where the precision is larger than the scale this actually becomes an error'

```
>>> df = spark.createDataFrame(data=[(Decimal('-1')), (Decimal('0'))], schema=DecimalType(7, 7))
>>> df.show()
20/11/25 12:52:57 ERROR Executor: Exception in task 0.0 in stage 32.0 (TID 125)
java.lang.ArithmeticException: Decimal precision 8 exceeds max precision 7
	at org.apache.spark.sql.types.Decimal.set(Decimal.scala:122)
	at org.apache.spark.sql.types.Decimal$.apply(Decimal.scala:571)
```

We cannot blindly add in -1 as a special case.

Reply:
Thanks for the exmaples. I have removed `-1` and `1` from special cases. Added another generator where scale and precision are same.